cpp/src/lance/arrow/dataset.cc

```diff
@@ -382,4 +391,52 @@ ::arrow::Result<::arrow::dataset::FragmentIterator> LanceDataset::GetFragmentsIm
   return ::arrow::MakeVectorIterator(fragments);
 }
 
+::arrow::Result<std::shared_ptr<LanceDataset>> LanceDataset::AddColumns(
+    const std::shared_ptr<::arrow::Table>& other,
+    const std::string& on,
+    ::arrow::MemoryPool* pool) {
+  /// Sanity checks
+  auto left_column = schema_->GetFieldByName(on);
+  if (left_column == nullptr) {
+    return ::arrow::Status::Invalid(fmt::format("Column {} does not exist in the dataset.", on));
+  }
+  auto right_column = other->GetColumnByName(on);
+  if (right_column == nullptr) {
+    return ::arrow::Status::Invalid(fmt::format("Column {} does not exist in the table.", on));
+  }
+
+  auto& left_type = left_column->type();
+  auto& right_type = right_column->type();
+  if (!left_type->Equals(right_type)) {
+    return ::arrow::Status::Invalid("LanceDataset::AddColumns: types are not equal: ",
+                                    left_type->ToString(),
+                                    " != ",
+                                    right_type->ToString());
+  }
+
+  // First phase, build hash table (in memory for simplicity)
+  auto merger = HashMerger(other, on, pool);
+  ARROW_RETURN_NOT_OK(merger.Init());
+
+  // Second phase
+  auto table_schema = other->schema();
+  ARROW_ASSIGN_OR_RAISE(auto merged_schema,
```

Comment:
`merged_schema` usually means the combined schema. It looks like this is the "schema that is being merged in" ?

Reply:
`schema_for_merging_columns`? 

any suggestions?

Reply:
Done